client/js/libs/handlebars/parse.js

```diff
@@ -1,126 +1,272 @@
-		text = parseArr[i];
-		match = text.match(back_re);
-		if (!match) {
-			// ^C (no colour) ending. Escape current colour and carry on
-			background = "";
-			continue;
+				result.push({
+					start: start,
+					end: end,
+					link: "http://" + text.slice(start, end)
+				});
+			}
+		} else if (channelPrefixes.indexOf(text[position]) !== -1 &&
+			(position === 0 || /[^0-9a-z]/i.test(text[position - 1]))) {
+			start = position;
+			end = position + 1;
+			while (end < text.length && /\S+/.test(text[end])) {
+				end += 1;
+			}
+			if (end - start > 1) {
+				result.push({
+					start: start,
+					end: end,
+					channel: text.slice(start, end)
+				});
+			}
 		}
-		colour = "irc-fg" + +match[1];
-		// set the background colour
-		if (match[3]) {
-			background = " irc-bg" + +match[3];
-		}
-		// update the parsed text result
-		result = result.replace(colourKey + text, styleTemplate({
-			style: colour + background,
-			text: text.slice(match[0].length)
-		}));
+		position += 1;
 	}
 
-	// Matching styles (italics/bold/underline)
-	// if only colours were this easy...
-	styles.forEach(function(style) {
-		if (result.indexOf(style.key) < 0) {
-			return;
+	// fill
+	position = 0;
+	for (var i = 0; i < result.length; i++) {
+		var textSegment = result[i];
+		if (textSegment.start > position) {
+			result.splice(i, 0, {
+				start: position,
+				end: textSegment.start
+			});
 		}
+		position = textSegment.end;
+	}
+	if (position < text.length) {
+		result.push({
+			start: position,
+			end: text.length
+		});
+	}
+
+	return result;
+}
+
+function anyIntersection(a, b) {
+	return a.start <= b.start && b.start <= a.end ||
+		a.start <= b.end && b.end <= a.end ||
+		b.start <= a.start && a.start <= b.end ||
+		b.start <= a.end && a.end <= b.end;
+}
+
+function fullProcess(text) {
+	var fragments = parseStyle(text).map(function(fragment, i, array) {
+		fragment.start = i === 0 ? 0 : array[i - 1].end;
+		fragment.end = fragment.start + fragment.text.length;
+		return fragment;
+	});
+
+	var cleanText = fragments.map(function(fragment) {
+		return fragment.text;
+	}).join("");
 
-		result = result.replace(style.keyregex, function(matchedTrash, matchedText) {
-			return styleTemplate({
-				style: style.style,
-				text: matchedText
+	var textParts = analyseText(cleanText);
+
+	return textParts.map(function(textPart) {
+		textPart.fragments = fragments
+			.filter(function(fragment) {
+				return anyIntersection(textPart, fragment);
+			})
+			.map(function(fragment) {
+				var start = Math.max(fragment.start, textPart.start);
+				var end = Math.min(fragment.end, textPart.end);
+				return Object_assign({}, fragment, {
+					start: start,
+					end: end,
+					text: fragment.text.slice(start - fragment.start, end - fragment.start)
+				});
 			});
-		});
+		return textPart;
 	});
 
-	return result;
 }
+
+Handlebars.registerHelper(
+	"parse", function(text) {
+		function createFragment(fragment) {
+			var className = "";
```

Comment:
You could make this an array and then call `.join` on it, this would fix the leading space.